src/db/read.rs

```diff
@@ -79,11 +105,52 @@ impl<'a> Read<'a> {
         self.map.get(key)
     }
 
-    pub fn scan(&'a self, opts: super::ScanOptions<'a>) -> impl Iterator<Item = prolly::Entry<'a>> {
-        super::scan::scan(&self.map, opts)
+    pub async fn scan(
+        &'a self,
+        opts: super::ScanOptions<'a>,
+        callback: impl Fn(prolly::Entry<'_>),
+    ) -> Result<(), ScanError> {
+        use ScanError::*;
+        fn send<'a>(
+            map: &'a prolly::Map,
+            opts: super::ScanOptions<'a>,
+            callback: impl Fn(prolly::Entry<'_>),
+        ) {
+            let use_index = opts.index_name.is_some();
+            for mut thing in super::scan::scan(map, opts) {
+                // TODO: It would be nice to return either the primary or secondary key,
+                // but decoding it is kind of a bitch. Cannot return composite key because it
+                // isn't a valid string!
+                if use_index {
+                    thing = prolly::Entry {
+                        key: &[],
+                        val: thing.val,
+                    }
+                }
+                callback(thing);
+            }
+        }
+        match opts.index_name {
+            Some(name) => {
+                let idx = self
+                    .indexes
+                    .get(name)
+                    .ok_or_else(|| UnknownIndexName(name.to_string()))?;
+                let guard = idx.get_map(&self.dag_read).await.map_err(GetMapError)?;
+                send(guard.get_map(), opts, callback);
```

Comment:
Kind of strange that we do `get_map` and then `get_map` again?

Maybe on of them should be renamed?

Reply:
Heh yeah, didn't notice that. Before it was `to_map` but Clippy bitched at me because there's a convention that `to_*` have a particular signature.

Reply:
I'm not really sure what would be better here so leaving it.